primitives/core/src/bandersnatch.rs

```diff
@@ -0,0 +1,1001 @@
+		fn decode<R: codec::Input>(i: &mut R) -> Result<Self, codec::Error> {
+			let buf = <Box<[u8; RING_VRF_CONTEXT_PARAMS_SERIALIZED_LEN]>>::decode(i)?;
+			let kzg =
+				KZG::deserialize_compressed(buf.as_slice()).map_err(|_| "KZG decode error")?;
+			Ok(RingVrfContext(kzg))
+		}
+	}
+
+	impl MaxEncodedLen for RingVrfContext {
+		fn max_encoded_len() -> usize {
+			<[u8; RING_VRF_CONTEXT_PARAMS_SERIALIZED_LEN]>::max_encoded_len()
+		}
+	}
+
+	impl TypeInfo for RingVrfContext {
+		type Identity = [u8; RING_VRF_CONTEXT_PARAMS_SERIALIZED_LEN];
+
+		fn type_info() -> scale_info::Type {
+			Self::Identity::type_info()
+		}
+	}
+
+	/// Ring VRF signature.
+	#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+	pub struct RingVrfSignature {
+		/// VRF (pre)outputs.
+		pub outputs: VrfIosVec<VrfOutput>,
+		/// Pedersen VRF signature.
+		signature: [u8; PEDERSEN_SIGNATURE_SERIALIZED_LEN],
+		/// Ring proof.
+		ring_proof: [u8; RING_PROOF_SERIALIZED_LEN],
+	}
+
+	#[cfg(feature = "full_crypto")]
+	impl Pair {
+		/// Produce a ring-vrf signature.
+		///
+		/// The signature is valid if the signing [`Pair`] is part of the ring from which
+		/// the [`RingProver`] has been derived.
+		pub fn ring_vrf_sign(&self, data: &VrfSignData, prover: &RingProver) -> RingVrfSignature {
+			// Hack used because backend signature type is generic over the number of ios
+			// @burdges can we provide a Vec version in `bandersnatch_vrfs` crate?
+			match data.vrf_inputs.len() {
+				0 => self.ring_vrf_sign_gen::<0>(data, prover),
+				1 => self.ring_vrf_sign_gen::<1>(data, prover),
+				2 => self.ring_vrf_sign_gen::<2>(data, prover),
+				3 => self.ring_vrf_sign_gen::<3>(data, prover),
+				_ => unreachable!(),
+			}
+		}
+
+		fn ring_vrf_sign_gen<const N: usize>(
+			&self,
+			data: &VrfSignData,
+			prover: &RingProver,
+		) -> RingVrfSignature {
+			let ios: Vec<_> = data
+				.vrf_inputs
+				.iter()
+				.map(|i| self.0.clone().0.vrf_inout(i.0.clone()))
+				.collect();
+
+			let ring_signature: bandersnatch_vrfs::RingVrfSignature<N> =
+				self.0.sign_ring_vrf(data.transcript.clone(), ios.as_slice(), prover);
+
+			let outputs: Vec<_> = ring_signature.preoutputs.into_iter().map(VrfOutput).collect();
+			let outputs = VrfIosVec::truncate_from(outputs);
+
+			let mut signature = [0; PEDERSEN_SIGNATURE_SERIALIZED_LEN];
+			ring_signature
+				.signature
+				.serialize_compressed(signature.as_mut_slice())
+				.expect("ped-signature serialization can't fail");
+
+			let mut ring_proof = [0; RING_PROOF_SERIALIZED_LEN];
+			ring_signature
+				.ring_proof
+				.serialize_compressed(ring_proof.as_mut_slice())
+				.expect("ring-proof serialization can't fail");
+
+			RingVrfSignature { outputs, signature, ring_proof }
+		}
+	}
+
+	impl RingVrfSignature {
+		/// Verify a ring-vrf signature.
+		///
+		/// The signature is valid if has been produced by a member of the ring from which
+		/// the [`RingVerifier`] has been derived.
```

Comment:
[maybe](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute)?
```
#[must_use]
```

Reply:
Yeah I agree. For the moment I preferred to stick with the same interface provided by the `verify` of the other primitives. In a follow up is probably a good idea to change `bool` to `Result<(), ()>`